Add Case List Optimizations

mkangia opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR adds optimization options (Caching and Lazy loading) to be set for case list columns that support optimization (based on the format for the column). This setting is then used by CommCare app to provide better performance for heavier case lists.
This change is behind a new feature flag and would deprecate the existing "Cache and Index" feature flag eventually.

Once the feature flag is enabled and the app is configured to be on required commcare version (>2.56),
for each menu, 
the user can mark a setting to show optimization options. This is added to make the user decision explicit to add optimizations.

Under Settings tab for a module, in "Advanced" settings:
![image](https://github.com/user-attachments/assets/60e97843-eb3d-4b32-84fd-f6e3562506d9)

---
with help text
![image](https://github.com/user-attachments/assets/f608d4da-80db-402f-8ad9-ecc9d06ef1ba)

Once set, the user can see the optimizations options for columns in case list and `N/A` where not applicable (depending on the format chosen).

![image](https://github.com/user-attachments/assets/4e3dc6b1-2e41-4e9c-85b8-c420a6e57dc5)


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SC-4060

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`CASE_LIST_OPTIMIZATIONS`

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally and limited the impact of code changes to only when feature flag is enabled.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests for new changes.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-7470

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
add UI to set optimizations

mkangia committed:
add new property to model

this would also save it on the model automatically when the object saved from the view

mkangia committed:
add optimizations to suite generation

mkangia committed:
add module setting to show/hide case list optimizations

mkangia committed:
show case list optimizations once set to show in settings

mkangia committed:
show optimizations only for specific formats

mkangia committed:
save optimizations only when supported

mkangia committed:
set optimization only when supported

mkangia committed:
pick val from ui for reliable selection

mkangia committed:
better naming

mkangia committed:
add optimizations on detail same as columns

mkangia committed:
reduce change impact radius

limit JS execution to only when applicable

mkangia committed:
include check for app support for optimizations when showing options

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/screen.js:
```diff
@@ -107,6 +107,10 @@ hqDefine("app_manager/js/details/screen", function () {
             return caseTileTemplate && featureFlag;
         });
+        self.showCaseListOptimizations = (
+            self.columnKey === 'short' &&
```
"short" is to refer to "case list".
"long" is used for "case detail".

This change is relevant only for case list.

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -1860,6 +1860,7 @@ class DetailColumn(IndexedSchema):
     field = StringProperty()
     useXpathExpression = BooleanProperty(default=False)
     format = StringProperty(exclude_if_none=True)
+    optimization = StringProperty(exclude_if_none=True)
```
this one saves the optimization on the app document when saved from UI.

mkangia commented on corehq/apps/app_manager/suite_xml/xml_models.py:
```diff
@@ -866,6 +866,8 @@ class Field(OrderedXmlObject):
     sort = StringField('@sort')
     print_id = StringField('@print-id')
+    lazy_loading = BooleanField('@lazy_loading', required=False)
+    cache_enabled = BooleanField('@cache_enabled', required=False)
```
this one is the xml object corresponding to each column used when generating suite file.

mkangia commented on corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:
```diff
@@ -267,6 +267,22 @@ <h4 class="panel-title panel-title-nolink">{% trans "Advanced" %}</h4>
+                  {% trans "Show Case List Optimizations" %}
+                  <span class="hq-help-template"
+                        data-title="{% trans "Case List Optimizations" %}"
+                        data-content="{% blocktrans %}Set caching and lazy loading for case list columns for better performance. Once saved, reload this page to reveal/hide relevant column on case list. {% endblocktrans %}"
```
Unfortunately, this needs a reload.

The way javascripts are aligned for "settings" tab and the details tabs ("case list" and "case detail"), I could not see a way to have this happen without a reload.

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -1921,12 +1922,18 @@ def from_json(cls, data):
                     item.value[lang] = interpolate_media_path(path)
+
+        if to_ret.optimization and not to_ret.supports_optimizations:
+            to_ret.optimization = None
```
just a stronger check on the model to read optimization from data from view only when supported.

mkangia replied:
Definitely not necessary.

It prevents any misconfiguration sent from the UI, if any. Just model validation to not rely on UI checks only.

>  what would break if the suite includes optimizations but they're not supported on the CC version?

So, this is a check for the format of the column. Optimization is supported only for specific ones.

The suite file has a check as well to only add optimizations when supported for the column.
In case it's incorrectly set for a format that does not support it, it might break the app or cause unexpected app behavior, because CC would try to optimize it, if the CC version is up to date.

mkangia commented on corehq/apps/app_manager/detail_screen.py:
```diff
@@ -311,7 +311,7 @@ def fields(self):
             # ones needed for multi sort, it's safe to assume that case list optimizations
             # should only be added only when multi sort is enabled
-            if self.app.supports_case_list_optimizations:
+            if self.app.supports_case_list_optimizations and self.column.supports_optimizations:
```
just a way to strengthen the check to add optimizations on the build files only when relevant.

mkangia replied:
Good point.

I am just thinking of an edge case where someone added the optimizations and then dropped the app version?

orangejenny replied:
hmm, yeah, unlikely but possible. Given that you explained that a misconfigured app will break on mobile, I support being cautious here.

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/column.js:
```diff
@@ -100,8 +100,12 @@ hqDefine("app_manager/js/details/column", function () {
         ];
-        self.optimizationSelectElement = uiElementSelect.new(optimizationOptions).val(self.original.optimization || "");
-        self.$optimizationSelectElement = $('<div/>').append(self.optimizationSelectElement.ui);
+        if (screen.showCaseListOptimizations) {
```
to limit the impact of this PR to only when applicable and reduce the impact radius.

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/screen.js:
```diff
@@ -110,7 +110,7 @@ hqDefine("app_manager/js/details/screen", function () {
         self.showCaseListOptimizations = (
             self.columnKey === 'short' &&
-            hqImport('hqwebapp/js/toggles').toggleEnabled('CASE_LIST_OPTIMIZATIONS') &&
+            initialPageData.get('app_supports_case_list_optimizations') &&
```
`app_supports_case_list_optimizations` includes the check for feature flag, so that was removed.

Charl1996 commented:
@mkangia Thanks for the comments along the way. It makes reviewing easier for someone without app manager experience (like myself).

Charl1996 commented on corehq/apps/app_manager/static/app_manager/js/details/column.js:
```diff
@@ -505,7 +505,10 @@ hqDefine("app_manager/js/details/column", function () {
+            // fetch val by searching the select element instead of directly from optimizationSelectElement since
+            // the event binding to fire change on optimizationSelectElement with change on select, gets lost
+            // when the element is hidden and then revealed again based on format.
+            column.optimization = self.supportsOptimizations() ? self.optimizationSelectElement.ui.find("select").val() : null;
```
Interesting

kaapstorm replied:
Maybe a different way to fix this is to set the event binding when the element is put (or put back) on the UI. I haven't checked this, but maybe lines 85-108 where the element is created could move down to about line 400 where the event binding is set.

Alternatively, instead of removing the element when an unsupported format is selected, maybe set its value to blank and disable it?

mkangia replied:
> to swap out the list of optimization options

I did try a few things and this was probably one of them. IIRC this didn't work well or got too complicated for me to pursue. I would definitely prefer to keep this as simple as possible since there is a lot already happening on this page and the number of JS files.

I am going to try the visible option.

> maybe set its value to blank and disable it?

Will give this a go as well.

> This sounds worth addressing

I am happy to revisit this in a follow up PR in case I am unable to improve on it here, to not block on the feature release. Since, it won't be in use for a while, till CC changes are released, we can make minor UI changes post this PR.

mkangia replied:
Visible binding worked!!

Turns out this would happen for any number of if blocks.

This fixed it.
https://github.com/dimagi/commcare-hq/pull/35755/commits/ac1a3b532ebd51945feee5011ff554cf7be9df02

The if block used is true/false on page load and does not change. And then we used visible binding for the two table data elements.

Thanks @orangejenny @kaapstorm

orangejenny replied:
excellent, glad to hear it

Charl1996 commented on corehq/apps/app_manager/suite_xml/sections/details.py:
```diff
@@ -592,6 +594,28 @@ def get_detail_id_safe(self, module, detail_type):
 
 
+def _add_detail_optimizations(module_detail, detail_xml_object):
+    """
```
So if I understand correctly, what is happening here is that "cache_enabled"/"lazy_loading" was previously (or currently) set on the detail by means of other feature flags, but now since we're adding the ability to add the optimisations in a more granular way we're just adding this method for backwards compatibility (or until the "other features" are being deprecated)?

mkangia replied:
Hey @Charl1996 
The product spec talks about how things are [currently implemented](https://docs.google.com/document/d/1LM2PBXqDrUOKVfk_p2niu_-D695goeAGLevZ0jrrMVc/edit?tab=t.0#heading=h.rocw784f8vlk). I am not certain if exactly the same properties are used though there is an overlap and functionally there are optimizations being done for case list via different feature flags, `CACHE_AND_INDEX` and `CASE_LIST_LAZY` (this one is for web apps).

> we're adding the ability to add the optimisations in a more granular way

That is true. We have given more control to the user to set what they want instead of internal deciding what should be done. Also, I believe, externally caching and lazy loading were not available as two separate options as they are now.

> or until the "other features" are being deprecated)

The timeline is to be decided but yes they would be deprecated and projects will be moved to this feature flag.

Charl1996 commented:
I'd be happy to approve if QA is done and all works as expected. I don't know the app-manager code well (or at all) so take my review with a pinch of salt.

Charl1996 reviewed (approved):
Good to go from my side since QA passed.

zandre-eng reviewed (approved):
Overall looks good from my side, just a few minor nits.

ajeety4 reviewed (approved):
Not quite familiar with the app manager codebase however the code changes looks quite clear. Appreciate the checks included at various places.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/modules/module_view_settings.html:
```diff
@@ -267,6 +267,22 @@ <h4 class="panel-title panel-title-nolink">{% trans "Advanced" %}</h4>
+                <label class="control-label {% css_label_class %}">
+                  {% trans "Show Case List Optimizations" %}
+                  <span class="hq-help-template"
+                        data-title="{% trans "Case List Optimizations" %}"
```
Are there docs on this feature to link to? This seems like the kind of feature where users are going to want a reference to exactly what the options mean.

mkangia replied:
Documents are yet to be created.

This is pretty early work to get things on HQ proactively. The mobile changes are yet to be done after which the documentation will be added and only once the new CC app is released, this would actually be a usable feature. So, documentation is on the way!

orangejenny commented on corehq/toggles/__init__.py:
```diff
@@ -1376,6 +1376,16 @@ def _ensure_search_index_is_enabled(domain, enabled):
+    TAG_CUSTOM,
+    namespaces=[NAMESPACE_DOMAIN],
+    # ToDo: Add help link when available
+    # help_link=''
```
Nitpick: Can you just create the page, even if it's blank for now? Then you don't have to come back to this later (and then you could link it in the help bubble, too).

mkangia replied:
Good point, checking with AEs if I can get a link in advance.

orangejenny commented on corehq/apps/app_manager/const.py:
```diff
@@ -103,3 +103,12 @@
 MOBILE_UCR_V1_ALL_REFERENCES = f"{MOBILE_UCR_V1_FIXTURE_PATTERN}|{MOBILE_UCR_V1_REFERENCES_PATTERN}"
 MOBILE_UCR_V1_CASE_LIST_REFERENCES_PATTERN = r"instance\('commcare:reports'\)/reports/report\[@id='.*'\]/rows/row"
+
+FORMATS_SUPPORTING_CASE_LIST_OPTIMIZATIONS = [
```
Although, now that I've read through the rest of the PR, I see you still need the list to pass to js, so maybe doing this wouldn't be an improvement.

mkangia replied:
Hey @orangejenny 

I had not considered that.

> I see you still need the list to pass to js

Ya. I could add it in classes and then set this constant, from the "template_form" on those classes. That would unblock on the js. Is that a better way to configure this?

mkangia replied:
Cool, I will park this for now.

kaapstorm reviewed (approved):
Nicely done. Nothing blocking, just a couple of thoughts.

mkangia commented on corehq/apps/app_manager/detail_screen.py:
```diff
@@ -307,6 +308,12 @@ def fields(self):
                 endpoint_action=self.action,
                 alt_text=self.alt_text,
             )
+            # since case list optimizations are only available for versions higher than the
```
Its basically in the context of this function, which returns different things based on different settings, the top one being `enable_multi_sort`.
This comment tries to highlight why optimizations are being added only in this first if block and not others. I thought anyone looking at it might wonder.
Since `enable_multi_sort` is also version dependent and is enabled for versions even older than one needed for optimizations, this change is only made in this if block here.

orangejenny replied:
oh, I see, I had missed the `if self.app.enable_multi_sort` above.

It's a small point, but I would move this code out of that block (so just reduce the indentation one line on this block). You're adding enough complexity to the code that you need a 3-line comment to explain, but all you're saving is a single additional version check.

mkangia commented on corehq/apps/app_manager/detail_screen.py:
```diff
@@ -329,6 +336,13 @@ def fields(self):
             )
 
+    @staticmethod
+    def _set_case_list_optimizations(field, optimization):
```
You are saying that I could set this directly when creating the field instance?
Yes.

I have done it this way to avoid fixing all tests that would fail since `lazy_loading` and `cache_enabled` do get added, if I set them on the instance directly, I believe with False/None but they don't if noting is set explicitly.
I thought it was an okay trade of effort. Should I do the hard way?

mkangia replied:
Added a draft PR to show what I meant, https://github.com/dimagi/commcare-hq/pull/35884/files

Locally I can see that tests start failing because of this difference in suite files
```
   -    <field>
   +    <field cache_enabled="false" lazy_loading="false">
```

I am inclined to keep it how it is here though I am happy to reconsider if that is better in a way.

orangejenny replied:
Ugh. I do think the code in the draft PR is easier to follow, but I understand not wanting to update all the tests.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_list_properties.html:
```diff
@@ -18,6 +18,9 @@
         <!-- ko ifnot: showCaseTileMappingColumn() -->
         <th class="col-sm-2">{% trans "Format" %}</th>
         <!-- /ko -->
+        <!-- ko if: showCaseListOptimizations -->
```
yeah - in particular, turn on the case tiles flag and try the "Don't use" option, the "Manually configure" option, and one of the templates:

![Screenshot 2025-03-04 at 11 58 27 AM](https://github.com/user-attachments/assets/82f01a38-eefd-4697-a094-f2c0639d8e53)

mkangia replied:
Thanks for pointing this @orangejenny 

The columns were getting incorrectly added for the column. So I moved the optimization column in the end instead of being in middle of case tile and mapping columns.

This is how it looks now,
![image](https://github.com/user-attachments/assets/4baa91ae-dd5e-4bd8-bd73-601601baab0b)

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/35755/commits/5ee1ac9ba4cc765ce95d626d9a7923a2dc84edbc

mkangia committed:
add help link

mkangia committed:
js code improvements

mkangia committed:
use better way to show blank option

mkangia committed:
refactor: remove optimizationsSupported

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/column.js:
```diff
@@ -244,6 +269,18 @@ hqDefine("app_manager/js/details/column", function () {
 
         self.format = uiElementSelect.new(menuOptions).val(self.original.format || null);
+        self.supportsOptimizations = ko.observable(false);
+        self.optimizationsSupported = function () {
```
I tried this.

Since the value for `supportsOptimizations` depends on `format` which is not an observable (`format` is a `uiElementSelect`), it does not update when format is changed.
Putting format as observable would need me to modify existing which I am hesitant about at this point.

Though I got rid of `optimizationsSupported`.

orangejenny replied:
okay, getting rid of `optimizationsSupported` is a nice simplification.

I'd like to get rid of those ui element widgets altogether, in part because they don't work smoothly with knockout. But that's way out of scope.

mkangia committed:
move optimization column to the end

- avoid breaking the complex setup for case tile mapping and config columns

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/column.js:
```diff
@@ -81,6 +81,31 @@ hqDefine("app_manager/js/details/column", function () {
+        let optimizationOptions = [
+            {
+                value: "",
+                label: "",
```
Cool, updated in commit, https://github.com/dimagi/commcare-hq/pull/35755/commits/15420b19ffe78c92e4c548955cfd34cf388887f1

mkangia commented on corehq/apps/app_manager/static/app_manager/js/modules/module_view.js:
```diff
@@ -173,6 +173,15 @@ hqDefine("app_manager/js/modules/module_view", function () {
             });
         }
 
+        var showCaseListOptimizationsElement = $('#show_case_list_optimization_options');
```
Thanks for pointing that out @zandre-eng 

Updated,
https://github.com/dimagi/commcare-hq/pull/35755/commits/235b15c1d1a5376dab971d697a8ef1cdcfb200af

mkangia commented on corehq/apps/app_manager/static/app_manager/js/modules/module_view.js:
```diff
@@ -173,6 +173,15 @@ hqDefine("app_manager/js/modules/module_view", function () {
         }
 
+        var showCaseListOptimizationsElement = $('#show_case_list_optimization_options');
+        if (showCaseListOptimizationsElement.length > 0) {
```
Cool, updated in commit, https://github.com/dimagi/commcare-hq/pull/35755/commits/235b15c1d1a5376dab971d697a8ef1cdcfb200af

mkangia committed:
use visible binding

- using if block to add/remove td resulted in $optimizationSelectElement losing its event bindings.
So instead, show/hide it.

Charl1996 reviewed (approved):
No concerns from my side. Thanks for updating!

mkangia commented:
Thanks @Charl1996 @kaapstorm @ajeety4 @orangejenny 

Merged this considering there is no blocking feedback and to unblock release this week. And I am happy to consider adding follow up PR(s) needed for anything you suggest I should change, so please feel free to add more comments/responses.
